src/server/game/Movement/MovementGenerators/ChaseMovementGenerator.cpp

```diff
@@ -0,0 +1,182 @@
+    if (!owner || !owner->IsAlive())
+        return false;
+
+    // our target might have gone away
+    Unit* const target = GetTarget();
+    if (!target)
+        return false;
+
+    // the owner might've selected a different target (feels like we shouldn't check this here...)
+    if (owner->GetVictim() != target)
+        return false;
+
+    // the owner might be unable to move (rooted or casting), pause movement
+    if (owner->HasUnitState(UNIT_STATE_NOT_MOVE) || owner->IsMovementPreventedByCasting())
+    {
+
+        owner->StopMoving();
+        return true;
+    }
+
+    bool  const mutualChase     = isMutualChase(owner, target);
+    float const hitboxSum       = owner->GetCombatReach() + target->GetCombatReach();
+    float const minRange        = _range ? _range->minRange + hitboxSum : CONTACT_DISTANCE;
+    float const minTarget       = (_range ? _range->minTolerance : 0.0f) + hitboxSum;
+    float const maxRange        = _range ? _range->maxRange + hitboxSum : owner->GetMeleeRange(target); // melee range already includes hitboxes
+    float const maxTarget       = _range ? _range->maxTolerance + hitboxSum : CONTACT_DISTANCE + hitboxSum;
+    Optional<ChaseAngle> angle  = mutualChase ? Optional<ChaseAngle>() : _angle;
+
+    // if we're already moving, periodically check if we're already in the expected range...
+    if (owner->HasUnitState(UNIT_STATE_CHASE_MOVE))
+    {
+        if (_rangeCheckTimer > diff)
+            _rangeCheckTimer -= diff;
+        else
+        {
+            _rangeCheckTimer = RANGE_CHECK_INTERVAL;
+            if (positionOkay(owner, target, _movingTowards ? NONE : minTarget, _movingTowards ? maxTarget : NONE, angle))
+            {
+                owner->StopMoving();
+                _path = nullptr;
+            }
+        }
+    }
+
+    // if we're done moving, we want to clean up
+    if (owner->HasUnitState(UNIT_STATE_CHASE_MOVE) && owner->movespline->Finalized())
+    {
+        _path = nullptr;
+        owner->ClearUnitState(UNIT_STATE_CHASE_MOVE);
+
+        // @todo is this really the right place to call ::Attack?
+        if (positionOkay(owner, target, minRange, maxRange, {}))
+        {
+            owner->SetInFront(target);
+            owner->Attack(target, true);
+        }
+    }
+
+    // if the target moved, we have to consider whether to adjust
+    if (_lastTargetPosition != target->GetPosition() || mutualChase != _mutualChase)
+    {
+        _lastTargetPosition = target->GetPosition();
+        _mutualChase = mutualChase;
+        if (owner->HasUnitState(UNIT_STATE_CHASE_MOVE) || !positionOkay(owner, target, minRange, maxRange, angle))
+        {
+            Creature* const cOwner = owner->ToCreature();
+            // can we get to the target?
+            if (cOwner && !target->isInAccessiblePlaceFor(cOwner))
+            {
+                cOwner->SetCannotReachTarget(true);
+                cOwner->StopMoving();
+                _path = nullptr;
```

Comment:
is it necessary to delete the previous pointer?

Reply:
`unique_ptr`